### ISSUE TITLE
Integrate MLX-DISTRIBUTED-TRAINING for Multi-Device Training

### ISSUE THREAD

**Issue by akaszubski** [akaszubski/realign#45]
## Overview
Integrate MLX-DISTRIBUTED-TRAINING to enable privacy-first distributed training across multiple Apple Silicon devices.

## Benefits
- **Train 1B+ parameter models** from scratch using consumer hardware
- **38.7% speedup** with 2 devices (based on benchmarks)
- **Privacy-preserving**: All training happens locally on your devices
- **MPI-based gradient synchronization** for efficient distributed training
- **Real-time monitoring dashboard** at `localhost:8050`
- **Advanced optimizations**: Flash Attention, GQA, RoPE, SwiGLU

## Technical Architecture
- **Data Parallelism**: Each device maintains full model, data is sharded
- **Gradient Averaging**: Synchronized using `mx.distributed.all_sum`
- **Memory Management**: Dynamic batching, gradient accumulation, activation checkpointing
- **Model**: 22-layer transformer with 2048 embedding dims, 16 attention heads

### 1. Environment Setup
- [ ] Install system dependencies: `brew install mpich`
- [ ] Clone MLX-DISTRIBUTED-TRAINING repository
- [ ] Create virtual environment and install dependencies
- [ ] Verify setup: `python scripts/verify_setup.py`

### 2. Network Configuration
- [ ] Setup SSH between devices
- [ ] Test network connectivity: `python scripts/test_network.py`
- [ ] Ensure 10Gbps network or Thunderbolt 4 connection
- [ ] Configure MPI hostfile

### 3. Training Infrastructure
- [ ] Implement `DistributedTrainer` integration
- [ ] Configure `TrainingConfig` for realign use case
- [ ] Setup gradient accumulation and memory management
- [ ] Enable mixed precision training
- [ ] Configure checkpointing system

### 4. Monitoring & Dashboard
- [ ] Setup real-time performance dashboard
- [ ] Configure metrics tracking (loss, throughput, memory)
- [ ] Add alerting for training issues
- [ ] Implement checkpoint recovery

### 5. Multi-Device Coordination
- [ ] Test primary/secondary device roles
- [ ] Verify gradient synchronization
- [ ] Benchmark training speed vs single device
- [ ] Optimize network communication

### 6. Testing & Validation
- [ ] Run minimal training example
- [ ] Verify gradient averaging correctness
- [ ] Test checkpoint save/restore
- [ ] Measure speedup with multiple devices

## Example Usage

```bash
# Primary device (Mac Studio)
./scripts/start_training.sh --role primary

# Secondary device (MacBook)
./scripts/start_training.sh --role secondary

# Monitor
open http://localhost:8050
```

```python
from src.training.distributed_trainer import DistributedTrainer, TrainingConfig
from src.models.unified_model import UnifiedModel

# Configuration
config = TrainingConfig(
    batch_size=16,
    gradient_accumulation_steps=4,
    max_memory_gb=32,
    mixed_precision=True,
    sync_weights_every=100
)

# Initialize model
model = UnifiedModel.from_pretrained("mlx-community/Llama-3.2-3B")

# Train across devices
trainer = DistributedTrainer(config=config, model=model)
await trainer.train_step(batch)
```

## Hardware Requirements
- macOS Sonoma 14.0+ (Apple Silicon)
- Python 3.11+
- MLX 0.20.0+
- 32GB+ RAM per device
- 10Gbps network or Thunderbolt 4
- SSH access configured

## Documentation
- [MLX Distributed Docs](https://ml-explore.github.io/mlx/build/html/usage/distributed.html)
- Local reference: `~/Documents/GitHub/distributed-mlx/mlx-disitrubted-training/`

## Labels
enhancement, mlx, distributed, training, priority:medium

**Comment by akaszubski** [comment]
## ✅ Superseded by Issue #48

This issue is now superseded by **#48: Delegate all training methods to mlx-lm-lora**

### Critical Discovery

**MLX already has distributed training built-in!** The external `mlx-distributed-training` package is unnecessary.

```python
import mlx.core as mx
import mlx.nn as nn

# Built into MLX core!
world = mx.distributed.init()
grads = nn.average_gradients(grads)
```

### Launch Distributed Training

```bash
# No external packages needed - built into MLX
mlx.launch --backend ring --hosts mac-pro,mac-studio \
  mlx_lm_lora.train --model llama-70b --train-mode dpo
```

### What We Discovered

- ✅ `mlx.distributed` is **built into MLX core** (no external packages needed)
- ✅ Ring backend optimized for Thunderbolt 5
- ✅ MPI backend for traditional HPC
- ✅ Works with mlx-lm-lora out of the box
- ✅ 30-40% speedup expected

### Why Not mlx-distributed-training?

The external package is redundant - MLX core already has everything we need:
- `mx.distributed.init()`
- `nn.average_gradients()`
- `mlx.launch` command
- Ring and MPI backends

### Documentation

- **ADR-002**: Complete distributed training guide
- **MLX_ECOSYSTEM_GUIDE.md**: MLX distributed reference
- **distributed_training_example.py**: Working example

### Recommendation

Close this issue as **duplicate** of #48. The mlx-lm-lora integration (Issue #48) provides:
- Distributed training via built-in mlx.distributed
- Advanced training methods (DPO, ORPO, GRPO, CPO)
- No external distributed packages needed
- Simpler architecture

See #48 for full implementation plan.

**Comment by akaszubski** [comment]
Closing as duplicate of #48. MLX core has built-in distributed training - external mlx-distributed-training package is unnecessary. See comments for details.
